### PR TITLE
Protect trees against missing feature values and missing labels

### LINKED ISSUE #393: Adaptive Random Forest Regressor/Hoeffding Tree Regressor splitting gives AttributeError
## Versions

**river version**: River 0.1.0
**Python version**:Python 3.7.1
**Operating system**: Windows 10 Enterprise v1803

## Describe the issue

Have been playing around with River/Creme for a couple of weeks now, and it's super useful for a project I am currently working on. That being said, I'm still pretty new to the workings of the algorithms, so unsure whether this is a bug or an issue with my setup.

When I call `learn_one` on the ARFR or HTR, I receive: "Attribute Error "NoneType" object has no attribute '_left'" from line 113 in `best_evaluated_split_suggestion.py`. 

I have implemented a delayed prequential evaluation algorithm, and inspecting the loop, the error seems to be thrown when the model after the first delay period has been exceeded - ie when the model can first make a prediction that isn't zero. Before this point, `learn_one` doesn't throw this error.

Currently, I am using the default ARFR as in the [example](https://riverml.xyz/latest/api/ensemble/AdaptiveRandomForestRegressor/#examples), with a simple Linear Regression as the leaf_model. The linear regression model itself has been working with my data, when not used in the ARFR. I want to try the ensemble/tree models with the data to see if accuracy is improved due to the drift detection methods that are included. 

Has anyone also seen this error being thrown, or know the causes of it? Let me know if more information is needed! Thanks.

### PR DESCRIPTION
**Pull request by smastelini**
Fix #393.

---
### TL;DR

- Protects Binary Search Tree-based attribute observers against the case where `None` is passed to trees' `learn_one`method and splits are attempted.
- It also makes the trees deal with the situation where a split feature's value is not found during the learning process.

---

Currently, the trees manipulate their branches by identifying them with integers. Why so? Well, this gives us a standardized way of dealing with numerical and categorical features. For instance, suppose a categorical feature carries a mixture of strings and numbers. If we identify the tree branches by their actual values, they will become a mess. Right?

Instead, we use increasing integer values starting from `0` till `n_branches - 1`. We also reserve negative values for special cases:
- `-1`: a branch was not found while sorting an instance down the tree
- `-999`: the current branch corresponds to an alternate tree of the Hoeffding Adaptive Tree (HAT) classifier/regressor. This code is the same one used in the original MOA implementation of HATC.

That said, this PR concerns the specific case where `-1` is returned while sorting an instance in the trees. If, while traversing the tree, we get `-1` as the branch identifier for the next node, that could mean two things:

1. The current split feature is nominal, and the category in the new incoming instance `x` was never seen before this point. Hence, the tree does not know what to do with the instance from this point onward.
2. The node's split feature value in `x` is missing.

The solution for (1) was already in place since the original code of `skmultiflow`:

- if a new category arises while learning, we create a new leaf to accommodate it :)
- if a new category arises while predicting, we use the last accessible node to make the predictions. In other words, that poor decision node whose branches do not encompass the new emerging value.

Okie Doki. But what about (2)? Well, since the original Hoeffding Tree framework does not tells us what to do, and the current implementation might break depending on the situation:

**Changed (keeping the following paragraph for discussion history tracking purposes):**
- ~if the missing value arises while updating the tree, do nothing. That's the point: we do not know which of the two branches representing the tests `<=` and `>` the instance should be passed on. Thus, this PR makes the learning process simply stopping there. This strategy impacts the trees at different levels: HAT* implementations and EFDT update all the nodes in the path (they leverage the information from the incomplete instance). The vanilla HTs do not, so they will ultimately ignore this example as it cannot reach a leaf. As discussed in #393, other strategies might be pursued in the future.~
- if the missing value happens at prediction time, we use the last accessible node to obtain the responses.

### CODE REVIEW

**Comment by jacobmontiel** [comment]
Hi @smastelini 

Could you provide a brief description of the changes. Code looks good and seems to be working, but what is the rationale behind such implementation :-)

**Comment by smastelini** [comment]
> Could you provide a brief description of the changes. Code looks good and seems to be working, but what is the rationale behind such implementation :-)

Sure! I'll update the description for future reference.

**Comment by jacobmontiel** [comment]
Thanks for the clear explanation @smastelini 

> if the missing value arises while updating the tree, do nothing. That's the point: we do not know which of the two branches representing the tests <= and > the instance should be passed on. Thus, this PR makes the learning process simply stopping there. 

The problem with stopping learning is that we are potentially ignoring a lot of data samples (depending on the application).  An extreme case would be a reading of a faulting sensor that stops working, and all the data from that point is ignored due to a single feature that is always missing. The alternative is to set a default path, 'go left' for example. This can be seen as imputing missing data with a constant value. Although it is a simplistic solution we get the benefit of learning from the sample, hopefully, the rest of the features will be useful. 

**Note:** If most of the features are missing then you will need more sophisticated imputation techniques. However, even with sophisticated techniques, there is a limit in the amount of missing data where the data where learning is still feasible.

I believe that we have to set an option in this case: 
- `treat_missing_as_constant=True` Set a default path to follow for all nodes if the feature value is missing.
- `treat_missing_as_constant=False` Do nothing, i.e. ignore this sample.
 
You could even argue that an option to handle missing data should be created as this is not part of the original algorithm.

**Comment by smastelini** [comment]
Nice point @jacobmontiel. I vote to always select a default branch to sort incoming instances with missing data. We can mention this design choice as a note in `BaseHoeffdingTree`. What do you think? The only missing piece is how to deal with missing categories at prediction time. 

IMHO, more sophisticated data imputation strategies should be implemented as transformer. They do not belong to the scope of the tree algorithms.

**Comment by MaxHalford** [comment]
How about going down the path that has seen the most samples?

**Comment by jacobmontiel** [comment]
> How about going down the path that has seen the most samples?

This is an option. However, for consistency, we might one to pick just a default branch to follow or even create a branch for a missing categorical value. 

What I mean by "consistency": Using the most common value is a valid imputation method. The equivalent for numeric features is to use the mean or median value. Since we are considering just a constant value for numeric data, we should be doing something similar for categorical data.

**Comment by smastelini** [comment]
Hi @jacobmontiel and @MaxHalford. Thanks for your suggestions. In the last commit, I enforced the following patterns to deal with missing decision features in new incoming instances:

- When learning: choose the subtree branch most traversed so far to pass the instance on (note that I use a `while` loop to cope with potential multiple missing features)
- When predicting: Use the last "reachable" decision node to provide responses rather than trying to sort the instance further down the tree.

I'll add a note in `BaseHoeffdingTree` explaining these design decisions.

File: river/tree/_base_tree.py [hunk 1]
```diff
@@ -68,6 +68,16 @@ class BaseHoeffdingTree(ABC):
...
         If True, enable merit-based tree pre-pruning.
+
+    Notes
+    -----
+    Hoeffding trees might face situations where an input feature previously used to make
+    a split decision is now absent in an incoming sample. In this case, the trees follow the
+    conventions:
+
+    - *Learning:* choose the subtree branch most traversed so far to pass the instance on.</br>
+    - * Predicting:* Use the last "reachable" decision node to provide responses.
```
Comment (smastelini) [inline river/tree/_base_tree.py:79]:
```suggestion
    - *Predicting:* Use the last "reachable" decision node to provide responses.
```

File: river/tree/_nodes/base.py [hunk 2]
```diff
@@ -187,6 +198,23 @@ def split_test(self) -> InstanceConditionalTest:
...
+        """Calculate the total weight seen by the node.
+
+        Returns
+        -------
+        Total weight seen.
+        """
+        w = 0.
+
+        for child in self._children.values():
+            if child is None:
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/tree/_nodes/base.py:212]: Here is a simpler (and faster) alternative:

```python
w = sum(filter(None, self._children.values()))

File: river/tree/_base_tree.py [hunk 3]
```diff
@@ -68,6 +68,16 @@ class BaseHoeffdingTree(ABC):
         If True, disable poor attributes to reduce memory usage.
     merit_preprune
         If True, enable merit-based tree pre-pruning.
+
+    Notes
+    -----
+    Hoeffding trees might face situations where an input feature previously used to make
+    a split decision is now absent in an incoming sample. In this case, the trees follow the
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_base_tree.py]: Agreed! Reworded in commit [5fdd3b8](https://github.com/online-ml/river/pull/394/commits/5fdd3b86ae7776c9c416f8419ab419f48d30cc97).

File: river/tree/_nodes/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/base.py]: Much more elegant this way! I changed this portion in commit Agreed! Reworded in commit [5fdd3b8](https://github.com/online-ml/river/pull/394/commits/5fdd3b86ae7776c9c416f8419ab419f48d30cc97).

File: river/tree/_nodes/hatc_nodes.py [hunk 4]
```diff
@@ -277,17 +277,33 @@ def learn_one(self, x, y, *, sample_weight=1., tree=None, parent=None, parent_br
...
+            # Instance contains a categorical value previously unseen by the split node
+            if self.split_test.max_branches() == -1 and split_feat in x:
+                # Creates a new learning node to encompass the new observed feature value
+                leaf_node = tree._new_learning_node(parent=self)
+                branch_id = self.split_test.add_new_branch(x[split_feat])
+                self.set_child(branch_id, leaf_node)
+                tree._n_active_leaves += 1
+                leaf_node.learn_one(x, y, sample_weight=sample_weight, tree=tree, parent=self,
+                                    parent_branch=branch_id)
+            # The split feature is not present in the instance. Hence, we pass the new example
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/hatc_nodes.py]: Reworded in commit [5fdd3b8](https://github.com/online-ml/river/pull/394/commits/5fdd3b86ae7776c9c416f8419ab419f48d30cc97).

File: river/tree/_nodes/hatr_nodes.py [hunk 5]
```diff
@@ -239,17 +239,33 @@ def learn_one(self, x, y, sample_weight, tree, parent, parent_branch):
...
+            # Instance contains a categorical value previously unseen by the split node
+            if self.split_test.max_branches() == -1 and split_feat in x:
+                # Creates a new learning node to encompass the new observed feature value
+                leaf_node = tree._new_learning_node(parent=self)
+                branch_id = self.split_test.add_new_branch(x[split_feat])
+                self.set_child(branch_id, leaf_node)
+                tree._n_active_leaves += 1
+                leaf_node.learn_one(x, y, sample_weight=sample_weight, tree=tree, parent=self,
+                                    parent_branch=branch_id)
+            # The split feature is not present in the instance. Hence, we pass the new example
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_nodes/hatr_nodes.py]: Reworded in commit [5fdd3b8](https://github.com/online-ml/river/pull/394/commits/5fdd3b86ae7776c9c416f8419ab419f48d30cc97).

File: river/tree/hoeffding_tree_classifier.py [hunk 6]
```diff
@@ -351,17 +351,46 @@ def learn_one(self, x, y, *, sample_weight=1.):
...
+            # Split node encountered a previously unseen categorical value (in a multi-way test),
+            # so there is no branch to sort the instance to
+            if current.split_test.max_branches() == -1 and split_feat in x:
+                # Creates a new branch to the new categorical value
+                leaf_node = self._new_learning_node(parent=current)
+                branch_id = current.split_test.add_new_branch(x[split_feat])
+                current.set_child(branch_id, leaf_node)
+                self._n_active_leaves += 1
+                leaf_node.learn_one(x, y, sample_weight=sample_weight, tree=self)
+            # The split feature is not present in the instance. Hence, we pass the new example
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/hoeffding_tree_classifier.py]: Reworded in commit [5fdd3b8](https://github.com/online-ml/river/pull/394/commits/5fdd3b86ae7776c9c416f8419ab419f48d30cc97).

File: river/tree/hoeffding_tree_regressor.py [hunk 7]
```diff
@@ -244,17 +244,46 @@ def learn_one(self, x, y, *, sample_weight=1.):
...
+            # Split node encountered a previously unseen categorical value (in a multi-way test),
+            # so there is no branch to sort the instance to
+            if current.split_test.max_branches() == -1 and split_feat in x:
+                # Creates a new branch to the new categorical value
+                leaf_node = self._new_learning_node(parent=current)
+                branch_id = current.split_test.add_new_branch(x[split_feat])
+                current.set_child(branch_id, leaf_node)
+                self._n_active_leaves += 1
+                leaf_node.learn_one(x, y, sample_weight=sample_weight, tree=self)
+            # The split feature is not present in the instance. Hence, we pass the new example
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/hoeffding_tree_regressor.py]: Reworded in commit [5fdd3b8](https://github.com/online-ml/river/pull/394/commits/5fdd3b86ae7776c9c416f8419ab419f48d30cc97).

File: river/tree/_attribute_observer/numeric_attribute_class_observer_binary_tree.py [hunk 8]
```diff
@@ -67,6 +67,8 @@ def probability_of_attribute_value_given_class(self, att_val, class_val):
         return 0.0
 
     def best_evaluated_split_suggestion(self, criterion, pre_split_dist, att_idx, binary_only):
+        if self._root is None:
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/tree/_attribute_observer/numeric_attribute_class_observer_binary_tree.py]: Good catch! Changed in commit [3b74fd9](https://github.com/online-ml/river/pull/394/commits/3b74fd93ece697b784ba39997bec599bbc05d4c5).

**Comment by smastelini** [comment]
Thanks for the comments, @jacobmontiel. The changes are in place.

Just as a reminder for future reference, the tests in #404 ought to increase the coverage reported to be decreasing here.

**Review by MaxHalford — APPROVED** [review]
LGTM!

**Review by jacobmontiel — APPROVED** [review]
Thanks fopr the last changes. The implementation is correct. The PR is ready to merge.
